File: backend/controllers/User.controller.go
package controllers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/go-redis/redis/v8"
	"github.com/jmoiron/sqlx"
	"github.com/lxndrrud/blog-project/models"
	"github.com/lxndrrud/blog-project/services"
)

type IUserController interface {
	GetUserAndPosts(ctx *gin.Context)
	LoginUser(ctx *gin.Context)
	RegisterUser(ctx *gin.Context)
}

func NewUserController(db *sqlx.DB, redisConn *redis.Client) IUserController {
	return &userController{
		userService: services.NewUserService(db, redisConn),
	}
}

type userController struct {
	userService interface {
		GetUserAndPosts(idUser int64) (models.User, []models.Post, models.IError)
		LoginUser(login, password string) (string, models.User, []models.Permission, models.IError)
		RegisterUser(login, password string) models.IError
	}
}

func (c userController) GetUserAndPosts(ctx *gin.Context) {
	if ctx.Query("idUser") == "" {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Не указан идентификатор пользователя!",
		})
		return
	}
	idUser, err := strconv.ParseInt(ctx.Query("idUser"), 10, 64)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Неверно указан идентификатор!",
		})
		return
	}
	user, posts, errService := c.userService.GetUserAndPosts(idUser)
	if errService != nil {
		ctx.JSON(errService.GetCode(), errService)
		return
	}
	ctx.JSON(http.StatusOK, map[string]any{
		"user":  user,
		"posts": posts,
	})
}

func (c userController) LoginUser(ctx *gin.Context) {
	var form struct {
		Login    string `json:"login"`
		Password string `json:"password"`
	}
	err := json.NewDecoder(ctx.Request.Body).Decode(&form)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Неверное тело запроса!",
		})
		return
	}
	ctx.Request.Body.Close()
	if form.Login == "" {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Не указан логин!",
		})
		return
	}
	if form.Password == "" {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Не указан пароль!",
		})
		return
	}
	token, user, permissions, errService := c.userService.LoginUser(form.Login, form.Password)
	if errService != nil {
		ctx.JSON(errService.GetCode(), map[string]any{
			"message": errService.GetMessage(),
		})
		return
	}
	ctx.JSON(http.StatusOK, map[string]any{
		"token":       token,
		"currentUser": user,
		"permissions": permissions,
	})
}

func (c userController) RegisterUser(ctx *gin.Context) {
	var form struct {
		Login    string `json:"login"`
		Password string `json:"password"`
	}
	err := json.NewDecoder(ctx.Request.Body).Decode(&form)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Неверное тело запроса!",
		})
		return
	}
	ctx.Request.Body.Close()
	if form.Login == "" {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Не указан логин!",
		})
		return
	}
	if form.Password == "" {
		ctx.JSON(http.StatusBadRequest, map[string]any{
			"message": "Не указан пароль!",
		})
		return
	}
	errService := c.userService.RegisterUser(form.Login, form.Password)
	if errService != nil {
		ctx.JSON(errService.GetCode(), map[string]any{
			"message": errService.GetMessage(),
		})
		return
	}
	ctx.JSON(http.StatusCreated, nil)
}

File: backend/src/controllers/User.controller.ts
import { Request, Response } from "express";
import { IUserService } from "../services/User.service";
import { IHttpErrorCreator } from "../utils/HttpErrorCreator";

export interface IUserController {
    getUserAndPosts(req: Request, res: Response): Promise<void>
    loginUser(req: Request, res: Response): Promise<void>
    registerUser(req: Request, res: Response): Promise<void>
}

export class UserController implements IUserController {
    constructor(
        private readonly userService: IUserService,
        private readonly errorCreator: IHttpErrorCreator
    ) {}

    public async getUserAndPosts(req: Request, res: Response) {
        const idUser = <string | undefined> req.query.idUser 
        if (!idUser) {
            return this.errorCreator.badRequest400(res, new Error('Не указан идентификатор пользователя!'))
        }
        const pIdUser = parseInt(idUser)
        try {
            const info = await this.userService.getUserAndPosts(pIdUser)
            res.status(200).json(info)
        } catch(error: any) {
            if (error instanceof Error) {
                this.errorCreator.internalServerError500(res, error)
            }
        }
    }

    public async loginUser(req: Request, res: Response) {
        let payload: {
            login: string,
            password: string
        }
        try {
            payload = {
                login: req.body.login,
                password: req.body.password
            }
        } catch(error: any) {
            return this.errorCreator.badRequest400(res, error)
        }
        try {
            const info = await this.userService.loginUser(payload.login, payload.password)
            res.status(200).json({
                token: info.session.token,
                currentUser: info.user,
                permissions: info.permissions
            })
        } catch(error: any) {
            if (error instanceof Error) {
                this.errorCreator.internalServerError500(res, error)
            }
        }
    }

    public async registerUser(req: Request, res: Response) {
        let payload: {
            login: string,
            password: string
        }
        try {
            payload = {
                login: req.body.login,
                password: req.body.password
            }
        } catch(error: any) {
            return this.errorCreator.badRequest400(res, error)
        }
        try {
            await this.userService.registerUser(payload.login, payload.password)
            res.status(201).end()
        } catch(error: any) {
            if (error instanceof Error) {
                this.errorCreator.internalServerError500(res, error)
            }
        }
    }

}
